Add symbolic link functionality for SMB2

GregBragg opened the pull request:
Fixes https://github.com/AgNO3/jcifs-ng/issues/245

Add symlink Error Context handling for SMB2

mbechler commented on src/test/java/jcifs/tests/SymLinkTest.java:
```diff
@@ -0,0 +1,188 @@
+                if (file.isDirectory()) {
+                    log.info("file.list() -> {}", Arrays.toString(file.list()));
+
+                    if (file.getName().equals("symlinkTestDir/")) {
```
These names could be specified in a configuration property, then the test also could be easily skipped when the setup was not done/symlinks are not supported.

GregBragg replied:
Same comment as below...

mbechler commented on src/test/java/jcifs/tests/SymLinkTest.java:
```diff
@@ -0,0 +1,188 @@
+
+    @Parameters ( name = "{0}" )
+    public static Collection<Object> configs () {
+        return getConfigs("smb1", "noUnicode", "forceUnicode", "noNTStatus", "noNTSmbs", "smb2", "smb30", "smb31");
```
As symlinks are only supported with SMB 3.1.1 I don't think there is a need to run this against SMB1 and other SMB2 versions.

GregBragg replied:
Admittedly I do not fully understand how you have your JUnit tests configured... I'll attempt to make a change.

mbechler replied:
Don't worry, I can take care of this when merging.

mbechler commented on src/test/java/jcifs/tests/SymLinkTest.java:
```diff
@@ -0,0 +1,188 @@
+    private SmbFile createSession() throws MalformedURLException, CIFSException {
+        Properties props = new Properties();
+        props.setProperty("jcifs.smb.client.minVersion", DialectVersion.SMB210.name());
+        props.setProperty("jcifs.smb.client.maxVersion", DialectVersion.SMB311.name());
```
The way I use the test configurations it would be better to have a version check as precondition (Assume) instead. And the minimum version should be 3.1.1.
Then the default context/utility functions can be used as well.

GregBragg replied:
Same comment as above...

mbechler reviewed (commented):
Sorry for the rather long delay, a bit of unexpected downtime...

Some comments inline.

GregBragg commented:
@mbechler Sorry for the lengthy delay... I'm going to start reviewing your feedback notes this week and next and address your comments. Thanks!

GregBragg commented on src/main/java/jcifs/internal/smb2/Smb2ErrorContextResponse.java:
```diff
@@ -0,0 +1,146 @@
+
+        // SymLinkErrorTag (4 bytes) (always 0x4C4D5953)
+        int symLinkErrorTag = SMBUtil.readInt4(buffer, bufferIndex);
+        if ( !Integer.toHexString(symLinkErrorTag).toUpperCase().equals("4C4D5953") ) {
```
According to the specs this is:
`SymLinkErrorTag (4 bytes): The server MUST set this field to 0x4C4D5953.`

I did not want to take a chance on comparing the native int to a value that I saw while debugging, I'd rather follow the spec explicitly.

mbechler replied:
I would think that symLinkErrorTag == 0x4C4D5953 would just work

GregBragg replied:
Yeah it works just not a fan of implicit conversions from hex to decimal

GregBragg commented on src/main/java/jcifs/internal/smb2/Smb2ErrorContextResponse.java:
```diff
@@ -0,0 +1,146 @@
+        // STATUS_STOPPED_ON_SYMLINK is always 0x00000000
+        this.errorId = SMBUtil.readInt4(buffer, bufferIndex);
+        if ( this.errorId != 0 ) {
+            throw new SMBProtocolDecodingException("ErrorId should be 0 for STATUS_STOPPED_ON_SYMLINK");
```
Yes I see your point... I didn't take into account the possibility of a `SMB2_ERROR_ID_SHARE_REDIRECT` ErrorId

GregBragg commented on src/main/java/jcifs/smb/SmbFile.java:
```diff
@@ -657,30 +659,72 @@ SmbFileHandleImpl openUnshared ( String uncPath, int flags, int access, int shar
+                            if (this.isFile() && !uncPath.endsWith("\\")) { // is this a file in a symlink directory?
+                                String fileName = uncPath.substring(uncPath.lastIndexOf("\\") + 1);
+                                log.debug("fileName -> {}", fileName);
+                                return openUnshared (realPath + fileName, flags, access, sharing, attrs, options);
```
As far as I have tested it does work for a symlink directory, I have been unit testing that use case but I will review.

GregBragg replied:
Works as designed for a symlink directory

mbechler replied:
I still believe we should construct the symlink target path according to the spec and not using a heuristic locating the share name inside the UNC paths (but afterwards check that the share is still the same).
As per "2.2.2.2.1.1 Handling the Symbolic Link Error Response" one should construct the unparsed path component (i.e. the path below the symlink) from the last _UnparsedPathLength_ bytes of the request UNC path and append that to the provided _SubstituteName_.

GregBragg replied:
I've been reading those specs over and over and I still can't see how your code is going to handle a target path for a Substitute Name such as `\??\D:\GWTEST\Finance\smbSource\subSmbSource\` in SmbFile for a symlink directory with the original request of `SmbFile:659 - openUnshared: \Finance\symlinkTestDir\ flags: 0 access: 1 attrs: 80 options: 0` without doing what I have already coded to extract the root path to resolve the target path.

In this case unparsedPathLength = 0, so what am I expected to extract for the unparsed path?

`2022-10-06 21:51:58 INFO  SymLinkTest:114 - Resource [symlinkTestDir/] is a directory.`
`2022-10-06 21:51:58 INFO  SymLinkTest:124 - Resource [symlinkTestDir/] is not a file.`
`2022-10-06 21:51:58 DEBUG SmbFile:852 - queryPath: \Finance\symlinkTestDir\`
`2022-10-06 21:51:58 INFO  SymLinkTest:128 - Resource [symlinkTestDir/] is a symbolic link.`
`2022-10-06 21:51:58 DEBUG SmbFile:659 - openUnshared: \Finance\symlinkTestDir\ flags: 0 access: 1 attrs: 80 options: 0`
`2022-10-06 21:51:58 DEBUG SmbFile:784 - SymLink Path -> \Finance\symlinkTestDir\`
`2022-10-06 21:51:58 INFO  Smb2ErrorDataFormat:78 - unparsedPathLength -> 0`
`2022-10-06 21:51:58 DEBUG SmbFile:789 - SymLink Length -> 196`
`2022-10-06 21:51:58 DEBUG SmbFile:791 - Absolute Path -> true`
`2022-10-06 21:51:58 DEBUG SmbFile:792 - Print Name -> D:\GWTEST\Finance\smbSource\subSmbSource\`
`2022-10-06 21:51:58 DEBUG SmbFile:795 - Substitute Name -> \??\D:\GWTEST\Finance\smbSource\subSmbSource\`
`2022-10-06 21:51:58 DEBUG SmbFile:801 - Root Path -> \Finance\smbSource\subSmbSource\`
`2022-10-06 21:51:58 DEBUG SmbFile:825 - Target Path -> \Finance\smbSource\subSmbSource\`
`2022-10-06 21:51:58 DEBUG SmbFile:659 - openUnshared: \Finance\smbSource\subSmbSource\ flags: 0 access: 1 attrs: 80 options: 0`

mbechler replied:
Judging from the first clarification section in 2.2.2.2.1.1, I don't think that an absolute symlink to a local drive should be resolved at all in that way:

> An absolute symbolic link located on the server links "\\MachX\ShareY\Public\ProtocolDocs" to "\??\D:\DonHall\MiscDocuments\PDocs".
> The original open request is for "\\MachX\ShareY\Public\ProtocolDocs\DailyDocs\[MS-SMB].doc".
> The unparsed portion of the path name will be "\DailyDocs\[MS-SMB].doc". Appending the substitute name with the unparsed portion of the file name gives the new target path name of "\??\D:\DonHall\MiscDocuments\PDocs\DailyDocs\[MS-SMB].doc".

Therefore that symlink would point to the local D: drive _on the client_ (which I don't think we would want to handle at all), not transparently to another server-side resource. A different story might be a absolute UNC path, which might point to the same share (this could be easily handled), or to another share or server (this would be much more difficult, as mentioned before)

GregBragg commented on src/main/java/jcifs/internal/smb2/Smb2ErrorContextResponse.java:
```diff
@@ -0,0 +1,146 @@
+        // PathBuffer (variable), print name that also identifies the target of the symbolic link
+        this.printName = Strings.fromUNIBytes(buffer, printNameOffset + bufferIndex, printNameLength);
+
+        return symLinkLength;
```
Again this will be complex logic and IMHO I don't believe it brings much value here, however if you have a suggested way of doing it then I'm all for it.

GregBragg commented on src/main/java/jcifs/smb/DirFileEntryEnumIterator2.java:
```diff
@@ -106,6 +116,30 @@ protected FileEntry open () throws CIFSException {
+
+                    int i = substituteName.indexOf(getParent().getLocator().getShare());
+                    String realPath = substituteName.substring(i + getParent().getLocator().getShare().length());
+                    log.debug("symLinkPath -> {}", uncPath);
```
Hmmm... you given me a lot to think about... selfishly I was only thinking of my use cases for symlinks on the same file server share, which are not relative to another file server share. Do we really need to support relative symlinks now?

However reviewing this again today I believe the code will support both as written?
https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/a8da655c-8b0b-415a-b726-16dc33fa5827

Here the debug output for a directory which is a symlink, the code only looks at the end of the symlink path, it doesn't care what the path looks at the start:
`2022-08-19 10:46:30 INFO  SymLinkTest:119 - Resource [symlinkTestDir/] is a symbolic link.
2022-08-19 10:46:30 DEBUG SmbFile:654 - openUnshared: \Finance\symlinkTestDir\ flags: 0 access: 1 attrs: 80 options: 0
2022-08-19 10:46:31 DEBUG SmbFile:700 - symLinkLength -> 196
2022-08-19 10:46:31 DEBUG SmbFile:703 - substituteName -> \??\D:\GWTEST\Finance\smbSource\subSmbSource\
2022-08-19 10:46:31 DEBUG SmbFile:707 - symLinkPath -> \Finance\symlinkTestDir\
2022-08-19 10:46:31 DEBUG SmbFile:708 - realPath -> \Finance\smbSource\subSmbSource\
2022-08-19 10:46:31 DEBUG SmbFile:654 - openUnshared: \Finance\smbSource\subSmbSource\ flags: 0 access: 1 attrs: 80 options: 0
2022-08-19 10:46:31 DEBUG DirFileEntryEnumIterator2:126 - symLinkLength -> 196
2022-08-19 10:46:31 DEBUG DirFileEntryEnumIterator2:129 - substituteName -> \??\D:\GWTEST\Finance\smbSource\subSmbSource\
2022-08-19 10:46:31 DEBUG DirFileEntryEnumIterator2:133 - symLinkPath -> \Finance\symlinkTestDir\
2022-08-19 10:46:31 DEBUG DirFileEntryEnumIterator2:134 - realPath -> \Finance\smbSource\subSmbSource\
2022-08-19 10:46:31 INFO  SymLinkTest:130 - file.list() -> [subSmbSource1.txt, subSmbSource2.txt]`

Having said all that I do not have detailed knowledge of all the use cases that your code supports, I'll probably need your help figuring this all out...

GregBragg replied:
@mbechler still working on this to support relative symlinks, need to unit test some more so I catch all scenarios. Hold off for now on your review, Thanks!

GregBragg replied:
I've updated the code to support relative symlinks, symlinks to targets on a remote file share I did not implement.

GregBragg commented on src/main/java/jcifs/smb/DirFileEntryEnumIterator2.java:
```diff
@@ -106,6 +116,30 @@ protected FileEntry open () throws CIFSException {
+                    log.debug("symLinkPath -> {}", uncPath);
+                    log.debug("realPath -> {}", realPath);
+
+                    return open(realPath);
```
Normally I don't put a static counter in a method to to control the recursion depth, as it can be problematic, however I found this on SO that appears to be thread safe, not sure if you agree with this though... the one marked as best answer appears to satisfy the requirement. What is your opinion?

https://stackoverflow.com/questions/53729546/set-maximum-recursion-depth-in-java

mbechler replied:
I wasn't thinking about a static variable, this will definitly cause concurrency issues, but more about another parameter for the open method that tracks the current depth and a limit based on that.

GregBragg replied:
What has been implemented based on the SO conversation is more robust than having to track a counter whether it be in a parameter to the method or a static counter. It's thread safe and works effectively, you can test it by changing the `maxLevel` to a lower number.

GregBragg commented on src/main/java/jcifs/smb/SmbFile.java:
```diff
@@ -657,30 +659,72 @@ SmbFileHandleImpl openUnshared ( String uncPath, int flags, int access, int shar
+                                return openUnshared (realPath + fileName, flags, access, sharing, attrs, options);
+                            }
+                            else {
+                                return openUnshared (realPath, flags, access, sharing, attrs, options);
```
Same feedback as above...

GregBragg commented on src/main/java/jcifs/smb/DirFileEntryEnumIterator2.java:
```diff
@@ -106,6 +116,30 @@ protected FileEntry open () throws CIFSException {
+            if (cr != null && cr.isReceived() && cr.getStatus() == NtStatus.NT_STATUS_STOPPED_ON_SYMLINK) {
+                try {
+                    byte[] errorData = cr.getErrorData();
+                    Smb2ErrorContextResponse ecr = new Smb2ErrorContextResponse();
```
As this logic is somewhat different in SmbFile I have chosen to create a private method `parseSymLinkErrorData` in each class that will be easier to maintain. The logic is not really suited to be implemented in a shareable class or method.

mbechler replied:
I'm not sure the logic actually needs to be different, if done according to the spec, as noted below.

GregBragg commented on src/main/java/jcifs/smb/SmbFile.java:
```diff
@@ -1106,6 +1161,16 @@ else if ( this.fileLocator.isRootOrShare() ) {
+        }
+        return false;
+    }
+
```
implemented...

GregBragg commented:
@mbechler I've addressed a lot of your feedback points, however I need your help and expertise on a few items that I might not have covered off completely. Thanks!

GregBragg commented:
@mbechler My apologies for the chatty commits over the last couple of weeks, however I am ready for another review. Thanks!

GregBragg commented:
@mbechler I rebased my branch to master to make sure they were in sync with the latest changes in master. Please review at your convenience and let me know if we can move forward with this PR. Thanks!

mbechler commented:
Thanks, unfortunately It's going to be another two to three weeks till I can get back on this.

Dirk-A commented:
Hi Greg, thanks for contribution to the jcifs SMB library, it stumbled our team a bit that no one else came across symbolic link issues much sooner ;-).  Anyway , we recently had issues with paths (Windows) containing symbolic links and it seemed your solution was the one to go to.  I tested the code from your branch for our specific needs and after 1 modification, it does the job very well.
Issue in class Smb2ErrorDataFormat.java : 
line 41:  8 bytes are read from the error buffer => it seems that on Windows Enterprise10/Server 2019, the symbolic link error tag is already read at this point and hence this.errorid already equals 4C4D5953 and not 0 => the SMBProtocolDecodingException is incorrectly thrown.

GregBragg commented:
> Hi Greg, thanks for contribution to the jcifs SMB library, it stumbled our team a bit that no one else came across symbolic link issues much sooner ;-). Anyway , we recently had issues with paths (Windows) containing symbolic links and it seemed your solution was the one to go to. I tested the code from your branch for our specific needs and after 1 modification, it does the job very well. Issue in class Smb2ErrorDataFormat.java : line 41: 8 bytes are read from the error buffer => it seems that on Windows Enterprise10/Server 2019, the symbolic link error tag is already read at this point and hence this.errorid already equals 4C4D5953 and not 0 => the SMBProtocolDecodingException is incorrectly thrown.

Interesting... according to the spec at https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/9ff8837c-c4f7-452b-9272-8818a119dae9 ErrorId for a symlink error should be zero `0x00000000` unless it's a share redirect. Please review the link and the class `Smb2ErrorContextResponse` for more context. The ErrorId is part of the 8 byte boundary.

I really don't think the Windows server version has anything to do with what you are seeing. I have to follow the spec that Microsoft has written so not sure what is going on... Can you attach a code snippit of your fix so that I can test it?

Dirk-A commented:
It seems to follow this spec;
https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/f15ae37d-a787-4bf2-9940-025a8c9c0022
Code snippet:
...
    	int symLinkLength = 0;
    	int symLinkErrorTag = 0;
        int bufferIndex = super.readErrorContextResponse(buffer);
        if ( this.errorId != 0 ) {
        	if (Integer.toHexString(this.errorId).toUpperCase().equals("4C4D5953")) {
        		symLinkErrorTag = this.errorId;
        		this.errorId = 0;
        		symLinkLength = this.errorDataLength;
	        } else {
	            throw new SMBProtocolDecodingException("ErrorId should be 0 for STATUS_STOPPED_ON_SYMLINK");
	        }
        } else {

	        // SymLinkLength (4 bytes)
	        symLinkLength = SMBUtil.readInt4(buffer, bufferIndex);
	        bufferIndex += 4;
	
	        // SymLinkErrorTag (4 bytes) (always 0x4C4D5953)
	        symLinkErrorTag = SMBUtil.readInt4(buffer, bufferIndex);
	        if ( !Integer.toHexString(symLinkErrorTag).toUpperCase().equals("4C4D5953") ) {
	            throw new SMBProtocolDecodingException("SymLinkErrorTag should be 0x4C4D5953");
	        }
	        bufferIndex += 4;
        }
....

Dirk-A commented:
It seems readErrorResponse in ServerMessageBlock2.java already strips of the first 8 bytes of the error and returns the actual ErrorData which should match the structure in https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/f15ae37d-a787-4bf2-9940-025a8c9c0022

GregBragg commented:
> It seems to follow this spec; https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/f15ae37d-a787-4bf2-9940-025a8c9c002
>

That's read after the 8 byte aligned boundary for the spec that I documented in my previous comment, it has to be read first before you resolve the symlink error data. I think you are the wring place in the byte buffer somehow.

GregBragg commented:
> Yes I did. As parseSymLinkErrorData (SmbFile) parses the actual ErrorData, should readSymLinkErrorResponse still call parent method readErrorContextResponse?

It has to because of the 8 byte aligned boundary according to spec https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/9ff8837c-c4f7-452b-9272-8818a119dae9

Dirk-A commented:
A directory symbolic link (towards folder subfolder 20210216) was created in M:\SymLink1 which on its turn is shared as folder \\IMPAX671WFM\SymLink1: 
M:\SymLink1\>mklink /D 20210216 \\Impax671wfm\20210216
=> tests were done on files while checking \\IMPAX671WFM\SymLink1\20210216

GregBragg commented:
> A directory symbolic link (towards folder subfolder 20210216) was created in M:\SymLink1 which on its turn is shared as folder \IMPAX671WFM\SymLink1:
> M:\SymLink1>mklink /D 20210216 \Impax671wfm\20210216
> => tests were done on files while checking \IMPAX671WFM\SymLink1\20210216

Ok so your share is a symbolic link directory? If so I did not code for that scenario.

Can you please provide your SMB config settings for the share (without disclosing any userid/password or private info to your share).

Dirk-A commented:
Hi Gregory,
We have our own SMB wrapper classes defined to control client pools and have operations separated between folders and files.  But I will share what I can:
                Properties p = new Properties();
	    	p.putAll(System.getProperties());

	        p.setProperty("jcifs.smb.client.soTimeout", "600000");
	        p.setProperty("jcifs.netbios.cachePolicy", "1200");
	        p.setProperty("jcifs.smb.client.dfs.disabled", "true");
	        p.setProperty("jcifs.smb.client.connTimeout","10000");
	        p.setProperty("jcifs.smb.client.responseTimeout","60000");
	    	
	    	//only support NTLM!
	    	if (authMethod.equalsIgnoreCase("NTLM")) 
	    	{
	    		authenticatedContext = new BaseContext(new PropertyConfiguration(p))
	    				.withCredentials(new NtlmPasswordAuthenticator(domain, user, password));
	    	}
			
			InetAddress addr = InetAddress.getByName(host);
			
			this.transport = authenticatedContext.getTransportPool().
														getSmbTransport(authenticatedContext, 
																		new UniAddress(addr), port, true);

                 .....
The check here is to see whether a certain folder exists (M:\Symlink1\20210216\test), and this folder is part of a symbolic linked folder (M:\Symlink1\20210216), located in a shared folder (M:\Symlink1 => \\host\Symlink1):  mklink /D 20210216 \\host\20210216.  Hence M:\Symlink1\20210216\test would resolve towards \\host\20210216\test

try (SmbFile smbFolder = new SmbFile("smb://host/Symlink1/20210216/test",authenticatedContext);) {
	result = smbFolder.exists();
}
=> this will throw the symbolic link exception, but the error response differs 8 bytes due to the fact the SYM tag is not located at the expected position (as explained in the previous comments)

GregBragg commented:
To me it looks like you have configured the symbolic link directory as the root share directory which I did not unit test for that use case scenario, however can you try adding this to your properties config and let me know if it resolves the issue?

`p.setProperty("jcifs.smb.client.minVersion", DialectVersion.SMB311.name());
p.setProperty("jcifs.smb.client.maxVersion", DialectVersion.SMB311.name());`

mbechler commented:
I believe that  it's the subdirectory that is the symlink (mklink /D 20210216 \host\20210216). A share root as a symlink is a curious edge case however, that we might want to test, not really sure what should/would happen in that case.

It is however a link to another share, so I suppose ("ErrorId for a symlink error should be zero 0x00000000 unless it's a share redirect") that could be the issue. Implementing that, unfortunately, will open a whole new can of worms.

GregBragg commented:
These docs from Microsoft are a pita to follow... I found this gem in https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/d4da8b67-c180-47e3-ba7a-d24214ac4aaa

`ErrorData (variable): A variable-length data field that contains extended error information. If the ErrorContextCount field in the response is nonzero, this field MUST be formatted as a variable-length array of SMB2 ERROR Context structures as specified in section [2.2.2.1](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/9ff8837c-c4f7-452b-9272-8818a119dae9). Each SMB2 ERROR Context MUST start at an 8-byte aligned boundary relative to the start of the SMB2 ERROR Response. **Otherwise, it SHOULD[<4>](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/a64e55aa-1152-48e4-8206-edd96444e7f7#Appendix_A_4) be formatted as specified in section [2.2.2.2](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/7722d25d-cec7-4baf-8343-9ef84f48a52c)**.`

So is there situations where there is not an 8 byte aligned boundary at the start of the error data? That I don't know yet...

So going by your statement @mbechler I tend to agree that if this is really a share redirect scenario then what has been coded will not support it and I really don't want to open that can of worms.

@Dirk-A can you please confirm what kind of symlink this is, does it point to another share on a different host?

mbechler commented:
> So is there situations where there is not an 8 byte aligned boundary at the start of the error data? That I don't know yet...

I believe this is only the case for STATUS_BUFFER_TOO_SMALL when ErrorData is not formatted as the ErrorContext structure.

Dirk-A commented:
> 
Adding 
p.setProperty("jcifs.smb.client.minVersion", DialectVersion.SMB311.name()); 
p.setProperty("jcifs.smb.client.maxVersion", DialectVersion.SMB311.name());
to the properties, did solve the issue and errorId returned 0 as expected
Concerning the symbolic link:
This refers to the same host =>
M:\Symlink1\20210216 = \\host\Symlink1\20210216 with 20210216 is a symbolic link towards
M:\20210216 =  \\host\20210216

GregBragg commented:
> Adding
> p.setProperty("jcifs.smb.client.minVersion", DialectVersion.SMB311.name());
> p.setProperty("jcifs.smb.client.maxVersion", DialectVersion.SMB311.name());
> to the properties, did solve the issue and errorId returned 0 as expected
> Concerning the symbolic link:
> This refers to the same host =>
> M:\Symlink1\20210216 = \host\Symlink1\20210216 with 20210216 is a symbolic link towards
> M:\20210216 = \host\20210216

Awesome! I was hoping those settings would work because the MS specs we are following are only for the SMB 3.11 protocol.

mbechler commented on src/main/java/jcifs/internal/smb2/Smb2ErrorContextResponse.java:
```diff
@@ -0,0 +1,146 @@
+
+        // skip, not needed
+        bufferIndex += 4; // ReparseTag (4 bytes) (always 0xA000000C)
+        bufferIndex += 2; // ReparseDataLength (2 bytes)
```
As we have already seen with the < 3.1.1 issue, detecting an unexpected response an erring out early takes out quite a bit of guesswork when debugging the resuling issue, therefore I see some value in it.

GregBragg replied:
Updated...

mbechler commented:
Added some more review comments.

I've encountered something similar with NegotiateContexts, even though the spec states that the reserved fields should be zeroed, older clients still send some garbage data instead.  So we probably need to explicitly check the negotitated version (therefore make that available to the response parsing code) and only conditionally perform the parsing.

GregBragg commented:
> Added some more review comments.
> 
> I've encountered something similar with NegotiateContexts, even though the spec states that the reserved fields should be zeroed, older clients still send some garbage data instead. So we probably need to explicitly check the negotitated version (therefore make that available to the response parsing code) and only conditionally perform the parsing.

Yeah I was thinking the same thing, check the config to see if they have it set to SMB 3.1.1 before parsing the error data, if not SMB 3.1.1 throw an exception as usual for a symlink path.

@mbechler you didn't mark this as requesting changes, will you be taking on the rest of the enhancements that have been noted?

GregBragg commented:
@mbechler I've added the check for SMB 3.1.1 version and addressed some of your review comments, please review.

@Dirk-A do you want to pull down the latest changes from my branch and beta test them? Up to you, no pressure.

Thanks!

GregBragg commented:
Hi @mbechler it's been awhile since we last commented on the PR, anyway we can move forward with getting this approved/promoted to master? Thanks!

mbechler commented:
I still believe we should make sure that we get the behavior in https://github.com/AgNO3/jcifs-ng/pull/309#discussion_r990163884 right/according to spec. I was planning to check against the Microsoft implementation, but haven't gotten around to.

mbechler commented:
So with the setup on the server (2K19) as follows:
```
C:\test\symlinktests>mklink /d relative-dir ..\x-test
symbolic link created for relative-dir <<===>> ..\x-test
C:\test\symlinktests>mklink relative-file ..\x-test\foo.txt
symbolic link created for relative-file <<===>> ..\x-test\foo.txt
C:\test\symlinktests>mklink absolute-file C:\test\x-test\foo.txt
symbolic link created for absolute-file <<===>> C:\test\x-test\foo.txt
C:\test\symlinktests>mklink /d absolute-dir C:\test\x-test\
symbolic link created for absolute-dir <<===>> C:\test\x-test\
C:\test\symlinktests>mklink /d unc-dir \\w2k19-single-dc\test\x-test\
symbolic link created for unc-dir <<===>> \\w2k19-single-dc\test\x-test\
C:\test\symlinktests>mklink unc-file \\w2k19-single-dc\test\x-test\foo.txt
symbolic link created for unc-file <<===>> \\w2k19-single-dc\test\x-test\foo.txt
```

after enabling bidirectional local/remote symlink following on a Win10 machine (`fsutil behavior set SymlinkEvaluation L2L:1 R2R:1 L2R:1 R2L:1`) this comes out pretty much as I had expected/described:

```
PS C:\Users\user> Get-Content \\w2k19-single-dc\test\symlinktests\relative-dir\foo.txt
this is the remote file
[Explorer shows target: \\w2k19-single-dc\test\x-test\ for relative-dir]
PS C:\Users\user> Get-Content \\w2k19-single-dc\test\symlinktests\relative-file
this is the remote file
[Explorer shows target: \\w2k19-single-dc\test\x-test\foo.txt for relative-file]
PS C:\Users\user> Get-Content \\w2k19-single-dc\test\symlinktests\absolute-dir\foo.txt
this is the local file
[Explorer shows target: C:\test\x-test\ for absolute-dir]
PS C:\Users\user> Get-Content \\w2k19-single-dc\test\symlinktests\absolute-file
this is the local file
[Explorer shows target: C:\test\x-test\foo.txt for \absolute-file]
PS C:\Users\user> Get-Content \\w2k19-single-dc\test\symlinktests\unc-file
this is the remote file
[Explorer shows target: \\w2k19-single-dc\test\x-test\foo.txt for unc-file]
PS C:\Users\user> Get-Content \\w2k19-single-dc\test\symlinktests\unc-dir\foo.txt
this is the remote file
[Explorer shows target: \\w2k19-single-dc\test\x-test for unc-dir]
```

Relevant packet captures, link error data: [capture.zip](https://github.com/AgNO3/jcifs-ng/files/9945692/capture.zip)


So I think the correct behavior would be:

1. Process the Link data according to spec (UnparsedNameLength/SubstituteName)
2. Check that the result is an UNC path. If it is a local path, I don't think these should be handled in the library in any way. Throwing an exception containing the target path may be appropriate, to allow handling by the user
3. Check that the UNC path is on the same server/share (otherwise handling like for DFS would have to be implemented, making things much more complicated), else, throw an exception.
4. Follow symlink

GregBragg commented:
> after enabling bidirectional local/remote symlink following on a Win10 machine (`fsutil behavior set SymlinkEvaluation L2L:1 R2R:1 L2R:1 R2L:1`) this comes out pretty much as I had expected/described:

Can you explain why you would do this on the client? Or am I misinterpreting this test case which I never would have anticipated for?

mbechler commented:
That was just to investigate how Windows would interpret the symlinks. As I mentioned I also don't think we should resolve remote to local symlinks. What I mean is that with the current implementation we will end up resolving to invalid targets (e.g. imagine an absolute local path that contains the share name, or an UNC path that contains the share name in the hostname, or what happens if it does not - indexOf = -1). This may cause strange, difficult to understand errors which we should avoid and instead do it according to spec and then be able to properly detect which type it is and whether to handle it or bail out with an error.

GregBragg commented:
Ok understood... However I'm still at a lost to move forward with any other changes for this contribution. I'd appreciate if you could take what I have already proposed and make the changes as you see fit to work with the current implementation as I don't have the detailed level of knowledge of the interval workings of this code base that you do. Thanks!

Dirk-A commented:
Hi Greg,
We recently encountered an issue with the determination of the root path, more specifically in SmbFile.java, 804-806:
            int i = substituteName.indexOf(this.getShare());
            String rootPath = substituteName.substring(i + this.getShare().length());
            log.debug("Root Path -> {}", rootPath);
=> it can occur here that the substituteName does not contain the share name, hence the rootPath will be constructed from a wrong substituteName substring.  
In our case:
SymLink Path -> \srvwfm\symlink1\20210216
SymLink Length -> 124
Absolute Path -> true
Print Name -> \\srvwfm\20210216
Substitute Name -> \??\UNC\srvwfm\20210216
Share Name -> symlink100  (length = 10)
Root Path -> rvwfm\20210216  (starting substring: -1 + 10 => 9)

GregBragg commented:
> Hi Greg, We recently encountered an issue with the determination of the root path, more specifically in SmbFile.java, 804-806: int i = substituteName.indexOf(this.getShare()); String rootPath = substituteName.substring(i + this.getShare().length()); log.debug("Root Path -> {}", rootPath); => it can occur here that the substituteName does not contain the share name, hence the rootPath will be constructed from a wrong substituteName substring. In our case: SymLink Path -> \srvwfm\symlink1\20210216 SymLink Length -> 124 Absolute Path -> true Print Name -> \srvwfm\20210216 Substitute Name -> \??\UNC\srvwfm\20210216 Share Name -> symlink100 (length = 10) Root Path -> rvwfm\20210216 (starting substring: -1 + 10 => 9)
> 
> This is the case with and without forcing the dialect to SMB311

Yep ok, got it... I never unit tested with a UNC path (if that is what I am interpreting here), however now I get what @mbechler is concerned about, however still not sure how to code it properly as I don't have the detailed knowledge of the internal code base.

GregBragg commented:
@Dirk-A can you pull down the latest changes and review/test please? I've refactored the code to follow the specs for resolving symlinks and absolute target paths starting with `\??\D:\` will not be handled as per @mbechler. Thanks!

GregBragg commented:
> So I think the correct behavior would be:
> 
> 1. Process the Link data according to spec (UnparsedNameLength/SubstituteName)
> 2. Check that the result is an UNC path. If it is a local path, I don't think these should be handled in the library in any way. Throwing an exception containing the target path may be appropriate, to allow handling by the user
> 3. Check that the UNC path is on the same server/share (otherwise handling like for DFS would have to be implemented, making things much more complicated), else, throw an exception.
> 4. Follow symlink

@mbechler Ok points 1, 2, 4 are completed. Point 3 is completed, however for DFS I have no internal knowledge of how this was implemented so I did not add any handling for this scenario.

Local target paths will not be resolved, this will be left up to the client to further resolve/process as they will know the share and share directory to construct the root path to the server's local symlink target which will be availble via `SmbFile.getSymLinkTargetPath()`.

Dirk-A commented:
Hi Gregory,
Here are my remarks after testing;
**SmbFile.java:**
withOpen method => as we use exist() checks on files/folders, the exception handling here should also take care of symbolic links:

Smb2CreateResponse createResp = cr.getResponse();
                            
                   if (createResp != null && createResp.isReceived() && createResp.getStatus() == NtStatus.NT_STATUS_STOPPED_ON_SYMLINK)

{                             this.isSymLink = true;                             Smb2SymLinkResolver resolver = new Smb2SymLinkResolver();                             String targetPath = resolver.parseSymLinkErrorData(createResp.getFileName(), createResp.getErrorData());                             this.symLinkTargetPath = targetPath;                             th.send(new Smb2CloseRequest(th.getConfig(), createResp.getFileId()), RequestParam.NO_RETRY);                    }
else if ( createResp.isReceived() && createResp.getStatus() == NtStatus.NT_STATUS_OK )

{                           th.send(new Smb2CloseRequest(th.getConfig(), createResp.getFileId()), RequestParam.NO_RETRY);                    }

**RecursionLimiter.java:**
As we use exist check + readable/writable checks, the recursion method stack can be larger than the default check on 50 => increase to 200

It still seems the intention is to only  support DialectVersion.SMB311, will this change in the near future?

Now only local paths are supported (\\??\\)
=> throws exception), but it is such an issue to take eg, the print name or the substitute name and strip off the ??\UNC part?
SymLink Path -> 20210216
Print Name -> \\hostname\20210216
Substitute Name -> ??\UNC\hostname\20210216
(as it should become)Target Path -> \hostname\20210216

mbechler commented:
Regarding 3.1.1: That is the only version that supports symlinks via this mechanism (providing the symlink data via error response). There might be an alternate route, e.g. making an IOCTL request if the error code is received, but have not looked at this at all.

Regarding absolute targets, if it's cross-share or cross-server things become a lot more complicate as now the requests need to be routed differently, and doing that transparently causes all kinds of issues, both architecture-wise and also logically (e.g. what credentials should be used in that case?). I haven't looked closely but I guess some improvement could be made regarding absolute targets on the same share, this should work as is, but I don't  think this applies to your example, which is cross-share.
An alternative option I see would be to provie the target as part of the exception, so that you could get a new SmbFile object for thfe target. Would that be helpful, too, or do you really need transparent handling?

GregBragg commented:
> An alternative option I see would be to provie the target as part of the exception, so that you could get a new SmbFile object for thfe target. Would that be helpful, too, or do you really need transparent handling?

The exception provided does supply the local target path as part of the message, however this can always be retrieved via `SmbFile.getSymLinkTargetPath()`.

Regarding `exist()` I believe we shouldn't resolve the path in there, that is more an architectural decision on my part to only identify that a symlink was encountered.

With regards to `RecursionLimiter` the check limit can certainly be increased but the question becomes what will be an ideal upper limit? Personally this was something I believe was not really needed, however it does limit run away threads in a loop executing forever. There are other libraries out there that don't limit recursion in any way and I have personally never had to implement anything to keep track or limit the recursion depth.

GregBragg commented:
@Dirk-A I tested your code suggestion:

> **SmbFile.java:** withOpen method => as we use exist() checks on files/folders, the exception handling here should also take care of symbolic links:
> 
> Smb2CreateResponse createResp = cr.getResponse(); if (createResp != null && createResp.isReceived() && createResp.getStatus() == NtStatus.NT_STATUS_STOPPED_ON_SYMLINK)
> 
> { this.isSymLink = true; Smb2SymLinkResolver resolver = new Smb2SymLinkResolver(); String targetPath = resolver.parseSymLinkErrorData(createResp.getFileName(), createResp.getErrorData()); this.symLinkTargetPath = targetPath; th.send(new Smb2CloseRequest(th.getConfig(), createResp.getFileId()), RequestParam.NO_RETRY); } else if ( createResp.isReceived() && createResp.getStatus() == NtStatus.NT_STATUS_OK )
> 
> { th.send(new Smb2CloseRequest(th.getConfig(), createResp.getFileId()), RequestParam.NO_RETRY); }

This line after resolving the symlink throws an exception trying to close the request:
`th.send(new Smb2CloseRequest(th.getConfig(), createResp.getFileId()), RequestParam.NO_RETRY);`

so I propose we could enhance it this way:
`Smb2CreateResponse createResp = cr.getResponse();`
`    if ( createResp != null && createResp.isReceived() && createResp.getStatus() == NtStatus.NT_STATUS_STOPPED_ON_SYMLINK ) {`
`        this.isSymLink = true;`
`        Smb2SymLinkResolver resolver = new Smb2SymLinkResolver();`
`        this.symLinkTargetPath = resolver.parseSymLinkErrorData(createResp.getFileName(), createResp.getErrorData());`
`    } else if ( createResp.isReceived() && createResp.getStatus() == NtStatus.NT_STATUS_OK ) {`
`        th.send(new Smb2CloseRequest(th.getConfig(), createResp.getFileId()), RequestParam.NO_RETRY);`
`}`

GregBragg commented:
@Dirk-A any feedback on the latest changes? It's been awhile...

@mbechler wondering when we can move forward with this or whether you need any other changes committed or whether you would like to take it from here and modify the contribution as you see fit...

Dirk-A commented:
Hi Gregory, 
bit busy here :-), next week I will plan to do some re-testing on the JCIFS library.

mbechler commented:
I should be able to take it from here. Hoping to be able to work on some stuff over christmas.

mbechler commented:
Have not found as much time as I would have liked, but I created a branch (https://github.com/AgNO3/jcifs-ng/tree/GregBragg-Add_symlink_functionality) where I made some changes (https://github.com/AgNO3/jcifs-ng/commit/fd5bd473d0da734157ea9edd8d2f3e130c10632f).  Still a work in progress though. I'm also hoping to implement symlink creation as well, as this would really be benefical for test cases.

mbechler commented:
Some more progress, implemented symlink creation etc, so I can work on automated test cases next. Also might have a solution for pre 3.1 versions.

GregBragg commented:
@mbechler It's been awhile since we had any activity on this enhancement. Any thoughts for when we can collaborate further on this PR proposal? Thanks!

mbechler commented:
My apologies, I haven't had much time lately and got a bit distracted. Hoping to get back to work soon.

Dirk-A commented:
Hi, it has indeed been a while.  My feedback is still that the limitation to MIN version SMB311 is not acceptable for our application needs. These are the local library changes I still maintain to cover all use cases we encountered in the field:

**Smb2ErrorDataFormat.java** =>

    public int readSymLinkErrorResponse ( byte[] buffer ) throws SMBProtocolDecodingException {
        int symLinkLength = 0;
    	int symLinkErrorTag = 0;
    	int bufferIndex = super.readErrorContextResponse(buffer);
        
        //DIRK: FIX TO GET RID OF DIALECT DEPENDENCY!
        if ( this.errorId != 0 ) {
        	if (Integer.toHexString(this.errorId).toUpperCase().equals("4C4D5953")) {
        		symLinkErrorTag = this.errorId;
        		this.errorId = 0;
        		symLinkLength = this.errorDataLength;
	        } else {
	        	throw new SMBProtocolDecodingException("ErrorId should be 0 for STATUS_STOPPED_ON_SYMLINK");
	        }
        } else {

	        // SymLinkLength (4 bytes)
	        symLinkLength = SMBUtil.readInt4(buffer, bufferIndex);
	        bufferIndex += 4;
	
	        // SymLinkErrorTag (4 bytes) (always 0x4C4D5953)
	        symLinkErrorTag = SMBUtil.readInt4(buffer, bufferIndex);
	        if ( symLinkErrorTag != 0x4C4D5953 ) {
	            throw new SMBProtocolDecodingException("SymLinkErrorTag should be 0x4C4D5953");
	        }
	        bufferIndex += 4;
        }
....

**SmbFile.java => remove throwing exceptions in case of :** 

1) if (config.getMinimumVersion() != DialectVersion.SMB311){ ....
2) if (targetPath.startsWith("\\??\\")) { ...
3)  and take share into account:
 this.symLinkTargetPath = resolver.parseSymLinkErrorData(createResp.getFileName(), createResp.getErrorData(),**this.getShare()**);

**Smb2SymLinkResolver.java => take share into account:**

        ....
        String rootPath = "";
        if (erdf.isAbsolutePath()) {
        	if ((null != shareName) && !shareName.isEmpty()) {
	        	log.debug("Share Name -> {}", shareName);
	            int i = substituteName.indexOf(shareName);
	            if (i > 0){
	            	rootPath = substituteName.substring(i + shareName.length());
	            } else {
	            	rootPath = erdf.getPrintName();
	            }
        	} else {
            	rootPath = erdf.getPrintName();
            }
        	log.debug("Root Path -> {}", rootPath);	
        	 if (!originalFileName.endsWith("\\") && rootPath.endsWith("\\")) { // is this a file in a symlink directory?
                 String fileName = originalFileName.substring(originalFileName.lastIndexOf("\\") + 1);
                 log.debug("File Name -> {}", fileName);
                 target = rootPath + fileName;
             }
             else {
                 target = rootPath + unparsedPath;
             }	
        } else {
        	 if (!originalFileName.endsWith("\\")) { // is this a file in a symlink directory?
                 String symLinkFileName = originalFileName.substring(originalFileName.lastIndexOf("\\") + 1);
                 log.debug("SymLink File Name -> {}", symLinkFileName);
                 target = originalFileName.replace(symLinkFileName, "") + substituteName;
             }
             else {
	            String parsedPath = getSymLinkParsedPath(originalFileName, unparsedPathLength);
	            StringBuilder b = new StringBuilder();
	            int startIndex = parsedPath.lastIndexOf("\\");
	            if (startIndex != -1) {
	                b.append(parsedPath, 0, startIndex);
	                b.append('\\');
	            }
	            b.append(substituteName);
	            b.append(unparsedPath);
	            target = b.toString();
             }
        }

mbechler commented:
Not entirely what I said. The symlink information is only available in the *error response* starting with 3.11, for earlier versions it has to be explicitly requested as part of the error handling. This need additional logic, but can be done. It will however be very inefficient, especially if the results are not cached.
